File: adapters/opencode.js
// OpenCode plugin — wires the existing skill picker into OpenCode's own
// plugin hooks. Reuses skills-launch.py verbatim for the "session start"
// side (feeding it a synthetic {"session_id": ...} stdin payload matching
// exactly what it already expects from Claude Code/Codex).
//
// Real API found by inspecting the installed @opencode-ai/plugin package
// directly (npm view + tarball extraction, 2026-07-22) rather than trusting
// the public docs alone — the docs' example plugin shape
// (`export const MyPlugin = async ({...}) => {{ ...hooks... }}`) turned out
// to be accurate for the CURRENT/default plugin API (a `Plugin =
// (input, options?) => Promise<Hooks>` function), which matches this file.
// There is also a newer, separate "v2" API (`define({id, setup(context)})`
// with a first-class `context.skill.transform()` for registering real skill
// sources) shipped in the same package under `./v2/promise` — not used here
// since the classic Hooks API is what's actually documented as the current
// plugin entry point; revisit if the v2 API becomes the primary one.
//
// Injection point: `chat.message` (appends a synthetic text Part to the
// incoming user message) — NOT `experimental.chat.system.transform`, which
// was tried first and, despite matching its documented type exactly,
// never actually fired at runtime across several real chat turns in live
// testing (2026-07-22) — no error, just never invoked. Plausible given the
// "experimental." prefix: declared in the types, not necessarily fully
// wired yet in this OpenCode version (v1.14.19). `chat.message` carries no
// such prefix and the SDK's `TextPart` type even has a dedicated
// `synthetic?: boolean` field for exactly this kind of injected content,
// so it's a better-supported mechanism, not just a fallback guess.

import { spawn } from "node:child_process";
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

const CACHE_DIR     = path.join(os.homedir(), ".claude", "cache");
const LAUNCH_SCRIPT  = path.join(os.homedir(), ".claude", "hooks", "skills-launch.py");
const CATALOG_PATH   = path.join(CACHE_DIR, "skills-catalog.json");

function readCatalog() {
  try {
    return JSON.parse(fs.readFileSync(CATALOG_PATH, "utf-8"));
  } catch {
    return [];
  }
}

/** Returns an array of skill instruction blocks for this session's picks,
 * or [] if the picker hasn't produced a pending file (yet, or at all). */
function resolvePicks(sessionID) {
  const pendingPath = path.join(CACHE_DIR, `skills-pending-${sessionID}.txt`);
  if (!fs.existsSync(pendingPath)) return [];

  let raw;
  try {
    raw = fs.readFileSync(pendingPath, "utf-8");
  } catch {
    return [];
  }
  try {
    fs.unlinkSync(pendingPath);
  } catch {
    // ignore — non-fatal, matches the other adapters' best-effort cleanup
  }

  const lines = raw.split("\n").map((l) => l.trim()).filter(Boolean);
  if (lines.length === 0) return [];

  const catalog = Object.fromEntries(readCatalog().map((e) => [e.name, e]));
  const blocks = [];
  for (const line of lines) {
    const sepIdx = line.indexOf("|");
    const name = (sepIdx === -1 ? line : line.slice(0, sepIdx)).trim();
    const arg  = (sepIdx === -1 ? "" : line.slice(sepIdx + 1)).trim();
    const entry = catalog[name];
    if (!entry) continue;

    let body = entry.description || "";
    if (entry.path && fs.existsSync(entry.path)) {
      try {
        body = fs.readFileSync(entry.path, "utf-8").trim();
      } catch {
        // fall back to description already assigned above
      }
    }
    if (!body) continue;

    const header = `### Skill: ${name}` + (arg ? ` (arg: ${arg})` : "");
    blocks.push(header + "\n\n" + body);
  }
  return blocks;
}

export const SkillPickerPlugin = async (_input) => {
  // Sessions already injected — NOT a cache of resolvePicks() results.
  // Real bug found live (2026-07-22): the very first chat.message call
  // fires ~100ms after session.created, before skills-launch.py/the picker
  // have had time to write a pending file — resolvePicks() correctly
  // returns [] for "not ready yet", but caching that empty result meant
  // the real pick (written seconds later, after the user actually
  // interacted with the picker) was never re-checked. Only mark a session
  // done AFTER a real successful injection.
  const injected = new Set();

  return {
    event: async ({ event }) => {
      if (event.type !== "session.created") return;
      const sessionID = event.properties.sessionID;
      try {
        const child = spawn("python3", [LAUNCH_SCRIPT], {
          stdio: ["pipe", "ignore", "ignore"],
          detached: true,
        });
        child.stdin.write(JSON.stringify({ session_id: sessionID }));
        child.stdin.end();
        child.unref();
      } catch {
        // best-effort — matches every other adapter's "never block/crash
        // the host on a picker-launch failure" rule
      }
    },

    "chat.message": async (input, output) => {
      const sessionID = input.sessionID;
      if (!sessionID || injected.has(sessionID)) return;

      const blocks = resolvePicks(sessionID);
      if (blocks.length === 0) return; // not ready yet (or nothing picked) — retry next turn

      const msg =
        "[STARTUP HOOK - The user selected these skills for this session. " +
        "Follow their instructions below as part of how you operate for the " +
        "rest of this session, starting now:\n\n" +
        blocks.join("\n\n---\n\n") +
        "]";

      try {
        // Part ids must start with "prt" — OpenCode's own schema validation
        // rejects anything else (confirmed via the exact SchemaError this
        // threw with our old "skillpicker-<sessionID>" id, which crashed
        // the whole message pipeline downstream, not just our hook).
        const partID = `prt_skillpicker${Date.now().toString(36)}${Math.random().toString(36).slice(2, 8)}`;
        output.parts.push({
          id: partID,
          sessionID,
          messageID: output.message.id,
          type: "text",
          text: msg,
          synthetic: true,
        });
      } catch {
        return; // don't mark as injected — retry next turn
      }

      // Inject once per session, not on every turn — matches the existing
      // Claude Code / Codex behavior (skills-inject.py / codex_inject.py).
      injected.add(sessionID);
    },
  };
};

export default SkillPickerPlugin;

File: adapters/opencode.py
#!/usr/bin/env python3
"""
OpenCode adapter — installs opencode.js (this same directory) as a global
OpenCode plugin at ~/.config/opencode/plugins/skill-picker.js.

Confirmed live against the actually-installed @opencode-ai/plugin package on
this machine (v1.14.19 in ~/.config/opencode/node_modules, checked directly
rather than trusting the public docs page, which describes a plugin export
shape that turned out to still be accurate — but only verified against the
real package's own type definitions and example plugin, since a second,
newer "v2" API also ships in the same package and is NOT what's used here).

Unlike Codex, OpenCode's extensibility is a real JS/TS plugin module, not a
shell-command hook config — there is no equivalent trust-approval gate to
navigate here (OpenCode loads plugins from ~/.config/opencode/plugins/
directly; no separate approval step is documented or was hit during
`adapters/codex.py`-style testing).

Usage:
    python3 adapters/opencode.py install
    python3 adapters/opencode.py uninstall
    python3 adapters/opencode.py status
"""

import os
import sys
import shutil

HOME              = os.path.expanduser("~")
OPENCODE_CONFIG   = os.path.join(HOME, ".config", "opencode")
PLUGINS_DIR       = os.path.join(OPENCODE_CONFIG, "plugins")
PLUGIN_DEST       = os.path.join(PLUGINS_DIR, "skill-picker.js")
ADAPTER_DIR       = os.path.dirname(os.path.abspath(__file__))
PLUGIN_SRC        = os.path.join(ADAPTER_DIR, "opencode.js")
CLAUDE_HOOKS_DIR  = os.path.join(HOME, ".claude", "hooks")
LAUNCH_SCRIPT     = os.path.join(CLAUDE_HOOKS_DIR, "skills-launch.py")

OPENCODE_APP_BUNDLE = "/Applications/OpenCode.app"


def is_installed():
    """Best-effort detection: bare `opencode` on PATH, an existing
    ~/.config/opencode/, or the desktop app bundle (confirmed present this
    way on at least one real machine)."""
    if shutil.which("opencode"):
        return True
    if os.path.isdir(OPENCODE_CONFIG):
        return True
    if os.path.exists(OPENCODE_APP_BUNDLE):
        return True
    return False


def install():
    if not is_installed():
        print("  ~ OpenCode not detected on this machine — skipping.")
        return False
    if not os.path.isfile(LAUNCH_SCRIPT):
        print(f"  x {LAUNCH_SCRIPT} not found — run this project's install.sh "
              f"first (the OpenCode adapter reuses skills-launch.py rather "
              f"than duplicating it).")
        return False
    if not os.path.isfile(PLUGIN_SRC):
        print(f"  x {PLUGIN_SRC} missing from the repo checkout.")
        return False

    os.makedirs(PLUGINS_DIR, exist_ok=True)
    shutil.copy(PLUGIN_SRC, PLUGIN_DEST)
    print(f"  + Installed {PLUGIN_DEST}")
    print("  ! Restart OpenCode (fully quit, not just close the window) for "
          "the new plugin to load, same as any other OpenCode plugin change.")
    return True


def uninstall():
    if os.path.isfile(PLUGIN_DEST):
        try:
            os.remove(PLUGIN_DEST)
            print(f"  + Removed {PLUGIN_DEST}")
        except OSError as e:
            print(f"  x Could not remove {PLUGIN_DEST}: {e}")
            return False
    else:
        print("  ~ No installed plugin found — nothing to remove.")
    return True


def status():
    installed = is_installed()
    present = os.path.isfile(PLUGIN_DEST)
    print(f"OpenCode detected: {installed}")
    print(f"Plugin installed: {present}")
    return installed and present


if __name__ == "__main__":
    cmd = sys.argv[1] if len(sys.argv) > 1 else ""
    if cmd == "install":
        sys.exit(0 if install() else 1)
    elif cmd == "uninstall":
        sys.exit(0 if uninstall() else 1)
    elif cmd == "status":
        sys.exit(0 if status() else 1)
    else:
        print(__doc__)
        sys.exit(1)
